##异常处理在工程化应用中是必须的操作命令，熟练掌握其应用技巧有助于写出完美的代码


### 1. 自定义异常

Python中所有的错误都继承与Exception类，所以我们也可以写自己定义的错误类
```python

class myerr(Exception):
    def __init__(self,err):
        Exception.__init__(self)
        self.err=err
try:
    raise myerr('myexception')
except myerr as var :
    print(var.err)
```
### 2. 异常处理

* 异常的抛出

```python
try:

     try块的语句...

except exceptiontype1 as var:#使用as语句获得本次捕获到的异常的实例  如果发生了第一种错误

　　except块语句...

except exceptiontype2 as var:#如果发生了第二种错误    可以做异常处理判断

　　except块语句...

except：

　　except块语句...

else:

　　else块语句...

finally:

　　finally块语句...

执行的流程分两类：


1.try->若有异常发生->except->finally
2.try->若无异常发生->else->finally

其中try块执行了可能抛出异常的语句，except块负责处理抛出的异常，处理的尝试顺序与多个except块的编写顺序一致，当尝试发现第一个异常类型匹配的except块时就进入该块内执行该except块的语句，最后一个不指定异常类型的except:  块匹配任何类型的异常
（由于所有异常都继承自Exception类所有except Exception：与不指定异常类型效果一致）。

```



